Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -627,4 +627,134 @@ public struct NonBlockingFileIO: Sendable {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            try Posix.unlink(pathname: path)
+        }
+    }
+
+    /// List contents of the directory at `path` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the directory to list the content of.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` containing the directory entries.
+    public func listDirectory(path: String, eventLoop: EventLoop) -> EventLoopFuture<[DirectoryEntry]> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            let dir = try Posix.opendir(pathname: path)
+            var entries: [DirectoryEntry] = []
+            do {
+                while let entry = try Posix.readdir(dir: dir) {
+                    let name = withUnsafeBytes(of: entry.pointee.d_name) { pointer -> String in
+                        let ptr = pointer.baseAddress!.assumingMemoryBound(to: CChar.self)
+                        return String(cString: ptr)
+                    }
+                    entries.append(DirectoryEntry(ino: UInt64(entry.pointee.d_ino), type: entry.pointee.d_type, name: name))
+                }
+                try? Posix.closedir(dir: dir)
+            } catch {
+                try? Posix.closedir(dir: dir)
+                throw error
+            }
+            return entries
+        }
+    }
+
+    /// Renames the file at `path` to `newName` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the file to be renamed.
+    ///     - newName: New file name.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` which is fulfilled if the rename was successful or fails on error.
+    public func rename(path: String, newName: String, eventLoop: EventLoop) -> EventLoopFuture<Void> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            try Posix.rename(pathname: path, newName: newName)
+        }
+    }
+
+    /// Removes the file at `path` on a private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// - parameters:
+    ///     - path: The path of the file to be removed.
+    ///     - eventLoop: The `EventLoop` on which the returned `EventLoopFuture` will fire.
+    /// - returns: An `EventLoopFuture` which is fulfilled if the remove was successful or fails on error.
+    public func remove(path: String, eventLoop: EventLoop) -> EventLoopFuture<Void> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            try Posix.remove(pathname: path)
+        }
+    }
+#endif
+}
+
+#if !os(Windows)
+/// A `DirectoryEntry` represents a single directory entry.
+public struct DirectoryEntry: Hashable {
```

Comment:
This needs to be namespaced with `NIO` in front of it.

Reply:
fixed, thanks!